File: compute-go/translator/strings.go
package translator

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

func translateGet(args []string) (Translation, error) {
	if len(args) != 1 {
		return Translation{}, errWrongArgs("get")
	}

	key := args[0]
	stmts := []Statement{
		deleteIfExpired(key),
		{
			SQL:    "SELECT type FROM keys WHERE key = ?",
			Params: []any{key},
		},
		{
			SQL:    "SELECT value FROM strings WHERE key = ?",
			Params: []any{key},
		},
	}

	return Translation{
		Statements: stmts,
		MapResult: func(results [][]map[string]any) (any, error) {
			if hasWrongType(results, 1) {
				return nil, ErrWrongType
			}
			if len(results[2]) == 0 {
				return nil, nil
			}
			return rowString(results[2][0], "value"), nil
		},
	}, nil
}

func translateSet(args []string) (Translation, error) {
	if len(args) < 2 {
		return Translation{}, errWrongArgs("set")
	}

	key := args[0]
	value := args[1]
	options, err := parseSetOptions(args[2:])
	if err != nil {
		return Translation{}, err
	}

	var expiresAt any
	if !options.keepTTL {
		if options.expiresAt != nil {
			expiresAt = *options.expiresAt
		} else {
			expiresAt = nil
		}
	}

	stmts := []Statement{deleteIfExpired(key)}
	getIndex := -1
	if options.get {
		getIndex = len(stmts)
		stmts = append(stmts, Statement{
			SQL:    "SELECT value FROM strings WHERE key = ?",
			Params: []any{key},
		})
	}

	okIndex := -1
	switch {
	case options.nx:
		okIndex = len(stmts)
		stmts = append(stmts,
			Statement{
				SQL:    "INSERT OR IGNORE INTO keys (key, type, expires_at) VALUES (?, 'string', ?) RETURNING key",
				Params: []any{key, expiresAt},
			},
			Statement{
				SQL:    "INSERT INTO strings (key, value) SELECT ?, ? WHERE changes() > 0",
				Params: []any{key, value},
			},
		)
	case options.xx:
		okIndex = len(stmts)
		if options.keepTTL {
			stmts = append(stmts, Statement{
				SQL:    "UPDATE keys SET type = 'string' WHERE key = ? RETURNING key",
				Params: []any{key},
			})
		} else {
			stmts = append(stmts, Statement{
				SQL:    "UPDATE keys SET type = 'string', expires_at = ? WHERE key = ? RETURNING key",
				Params: []any{expiresAt, key},
			})
		}
		stmts = append(stmts, Statement{
			SQL:    "INSERT OR REPLACE INTO strings (key, value) SELECT ?, ? WHERE changes() > 0",
			Params: []any{key, value},
		})
	default:
		if options.keepTTL {
			stmts = append(stmts,
				Statement{
					SQL:    "UPDATE keys SET type = 'string' WHERE key = ? RETURNING key",
					Params: []any{key},
				},
				Statement{
					SQL:    "INSERT INTO keys (key, type) SELECT ?, 'string' WHERE changes() = 0",
					Params: []any{key},
				},
				Statement{
					SQL:    "INSERT OR REPLACE INTO strings (key, value) VALUES (?, ?)",
					Params: []any{key, value},
				},
			)
		} else {
			stmts = append(stmts,
				Statement{
					SQL:    "DELETE FROM keys WHERE key = ?",
					Params: []any{key},
				},
				Statement{
					SQL:    "INSERT INTO keys (key, type, expires_at) VALUES (?, 'string', ?)",
					Params: []any{key, expiresAt},
				},
				Statement{
					SQL:    "INSERT INTO strings (key, value) VALUES (?, ?)",
					Params: []any{key, value},
				},
			)
		}
	}

	return Translation{
		Statements: stmts,
		MapResult: func(results [][]map[string]any) (any, error) {
			var oldValue any
			if getIndex >= 0 {
				if len(results[getIndex]) == 0 {
					oldValue = nil
				} else {
					oldValue = rowString(results[getIndex][0], "value")
				}
			}

			if okIndex >= 0 && len(results[okIndex]) == 0 {
				return nil, nil
			}
			if options.get {
				return oldValue, nil
			}
			return "OK", nil
		},
	}, nil
}

func translateDel(args []string) (Translation, error) {
	if len(args) == 0 {
		return Translation{}, errWrongArgs("del")
	}

	stmts := make([]Statement, 0, len(args))
	for _, key := range args {
		stmts = append(stmts, Statement{
			SQL:    "DELETE FROM keys WHERE key = ? RETURNING key",
			Params: []any{key},
		})
	}

	return Translation{
		Statements: stmts,
		MapResult: func(results [][]map[string]any) (any, error) {
			var count int64
			for _, res := range results {
				count += int64(len(res))
			}
			return count, nil
		},
	}, nil
}

func translateExists(args []string) (Translation, error) {
	if len(args) == 0 {
		return Translation{}, errWrongArgs("exists")
	}

	stmts := make([]Statement, 0, len(args))
	for _, key := range args {
		stmts = append(stmts, Statement{
			SQL:    "SELECT 1 FROM keys WHERE key = ? AND (expires_at IS NULL OR expires_at > unixepoch())",
			Params: []any{key},
		})
	}

	return Translation{
		Statements: stmts,
		MapResult: func(results [][]map[string]any) (any, error) {
			var count int64
			for _, res := range results {
				if len(res) > 0 {
					count++
				}
			}
			return count, nil
		},
	}, nil
}

func translateIncr(args []string) (Translation, error) {
	return translateIncrByDelta(args, 1, "incr")
}

func translateIncrBy(args []string) (Translation, error) {
	if len(args) != 2 {
		return Translation{}, errWrongArgs("incrby")
	}

	delta, ok := parseInt64(args[1])
	if !ok {
		return Translation{}, ErrNotInteger
	}
	return translateIncrByDelta(args[:1], delta, "incrby")
}

func translateDecr(args []string) (Translation, error) {
	return translateIncrByDelta(args, -1, "decr")
}

func translateDecrBy(args []string) (Translation, error) {
	if len(args) != 2 {
		return Translation{}, errWrongArgs("decrby")
	}

	delta, ok := parseInt64(args[1])
	if !ok {
		return Translation{}, ErrNotInteger
	}
	return translateIncrByDelta(args[:1], -delta, "decrby")
}

func translateIncrByDelta(args []string, delta int64, cmd string) (Translation, error) {
	if len(args) != 1 {
		return Translation{}, errWrongArgs(cmd)
	}

	key := args[0]
	stmts := []Statement{
		deleteIfExpired(key),
		{
			SQL:    "SELECT type FROM keys WHERE key = ?",
			Params: []any{key},
		},
		{
			SQL:    "INSERT OR IGNORE INTO keys (key, type) VALUES (?, 'string')",
			Params: []any{key},
		},
		{
			SQL:    "INSERT OR IGNORE INTO strings (key, value) VALUES (?, '0')",
			Params: []any{key},
		},
		{
			SQL:    "SELECT value FROM strings WHERE key = ?",
			Params: []any{key},
		},
		{
			SQL:    "UPDATE strings SET value = CAST(CAST(value AS INTEGER) + ? AS INTEGER) WHERE key = ? RETURNING value",
			Params: []any{delta, key},
		},
	}

	return Translation{
		Statements: stmts,
		MapResult: func(results [][]map[string]any) (any, error) {
			if hasWrongType(results, 1) {
				return nil, ErrWrongType
			}
			if len(results[4]) == 0 {
				return nil, ErrNotInteger
			}
			before := rowString(results[4][0], "value")
			if _, ok := parseInt64(before); !ok {
				return nil, ErrNotInteger
			}
			if len(results[5]) == 0 {
				return nil, ErrNotInteger
			}
			after := rowString(results[5][0], "value")
			value, ok := parseInt64(after)
			if !ok {
				return nil, ErrNotInteger
			}
			return value, nil
		},
	}, nil
}

func translateMGet(args []string) (Translation, error) {
	if len(args) == 0 {
		return Translation{}, errWrongArgs("mget")
	}

	stmts := make([]Statement, 0, len(args)*2)
	for _, key := range args {
		stmts = append(stmts,
			deleteIfExpired(key),
			Statement{
				SQL:    "SELECT value FROM strings WHERE key = ?",
				Params: []any{key},
			},
		)
	}

	return Translation{
		Statements: stmts,
		MapResult: func(results [][]map[string]any) (any, error) {
			out := make([]any, 0, len(args))
			for i := range args {
				res := results[i*2+1]
				if len(res) == 0 {
					out = append(out, nil)
					continue
				}
				out = append(out, rowString(res[0], "value"))
			}
			return out, nil
		},
	}, nil
}

func translateMSet(args []string) (Translation, error) {
	if len(args) == 0 || len(args)%2 != 0 {
		return Translation{}, errWrongArgs("mset")
	}

	stmts := make([]Statement, 0, (len(args)/2)*3)
	for i := 0; i < len(args); i += 2 {
		key := args[i]
		value := args[i+1]
		stmts = append(stmts,
			Statement{
				SQL:    "DELETE FROM keys WHERE key = ?",
				Params: []any{key},
			},
			Statement{
				SQL:    "INSERT INTO keys (key, type) VALUES (?, 'string')",
				Params: []any{key},
			},
			Statement{
				SQL:    "INSERT INTO strings (key, value) VALUES (?, ?)",
				Params: []any{key, value},
			},
		)
	}

	return Translation{
		Statements: stmts,
		MapResult: func(results [][]map[string]any) (any, error) {
			return "OK", nil
		},
	}, nil
}

type setOptions struct {
	nx        bool
	xx        bool
	get       bool
	keepTTL   bool
	expiresAt *int64
}

func parseSetOptions(args []string) (setOptions, error) {
	var opts setOptions
	var hasExpiry bool
	for i := 0; i < len(args); i++ {
		switch strings.ToUpper(args[i]) {
		case "NX":
			opts.nx = true
		case "XX":
			opts.xx = true
		case "GET":
			opts.get = true
		case "KEEPTTL":
			opts.keepTTL = true
		case "EX", "PX", "EXAT", "PXAT":
			if hasExpiry {
				return setOptions{}, errWrongArgs("set")
			}
			if i+1 >= len(args) {
				return setOptions{}, errWrongArgs("set")
			}
			value, ok := parseInt64(args[i+1])
			if !ok {
				return setOptions{}, ErrNotInteger
			}
			var expires int64
			switch strings.ToUpper(args[i]) {
			case "EX":
				expires = time.Now().Unix() + value
			case "PX":
				expires = time.Now().Unix() + value/1000
			case "EXAT":
				expires = value
			case "PXAT":
				expires = value / 1000
			}
			opts.expiresAt = &expires
			hasExpiry = true
			i++
		default:
			return setOptions{}, errWrongArgs("set")
		}
	}

	if opts.nx && opts.xx {
		return setOptions{}, errWrongArgs("set")
	}
	if opts.keepTTL && hasExpiry {
		return setOptions{}, errWrongArgs("set")
	}

	return opts, nil
}

func hasWrongType(results [][]map[string]any, index int) bool {
	return wrongTypeFor(results, index, "string")
}

func rowString(row map[string]any, key string) string {
	value, ok := row[key]
	if !ok || value == nil {
		return ""
	}
	switch v := value.(type) {
	case string:
		return v
	case []byte:
		return string(v)
	default:
		return fmt.Sprint(v)
	}
}

func parseInt64(value string) (int64, bool) {
	parsed, err := strconv.ParseInt(value, 10, 64)
	if err != nil {
		return 0, false
	}
	return parsed, true
}

File: compute-node/src/translator/strings.ts
import type { Statement, Translation, ResultRow } from "./types";
import { ErrNotInteger, ErrWrongType, errWrongArgs } from "./errors";
import { deleteIfExpired, parseIntStrict, rowString, wrongTypeFor } from "./helpers";

function hasWrongType(results: ResultRow[][], index: number): boolean {
  return wrongTypeFor(results, index, "string");
}

type SetOptions = {
  nx: boolean;
  xx: boolean;
  get: boolean;
  keepTTL: boolean;
  expiresAt: number | null;
};

function parseSetOptions(args: string[]): SetOptions {
  const options: SetOptions = {
    nx: false,
    xx: false,
    get: false,
    keepTTL: false,
    expiresAt: null,
  };
  let hasExpiry = false;

  for (let i = 0; i < args.length; i++) {
    const token = args[i].toUpperCase();
    switch (token) {
      case "NX":
        options.nx = true;
        break;
      case "XX":
        options.xx = true;
        break;
      case "GET":
        options.get = true;
        break;
      case "KEEPTTL":
        options.keepTTL = true;
        break;
      case "EX":
      case "PX":
      case "EXAT":
      case "PXAT": {
        if (hasExpiry) {
          throw errWrongArgs("set");
        }
        if (i + 1 >= args.length) {
          throw errWrongArgs("set");
        }
        const parsed = parseIntStrict(args[i + 1]);
        if (parsed === null) {
          throw ErrNotInteger;
        }
        const now = Math.floor(Date.now() / 1000);
        let expiresAt: number;
        switch (token) {
          case "EX":
            expiresAt = now + parsed;
            break;
          case "PX":
            expiresAt = now + Math.trunc(parsed / 1000);
            break;
          case "EXAT":
            expiresAt = parsed;
            break;
          case "PXAT":
            expiresAt = Math.trunc(parsed / 1000);
            break;
          default:
            throw errWrongArgs("set");
        }
        options.expiresAt = expiresAt;
        hasExpiry = true;
        i++;
        break;
      }
      default:
        throw errWrongArgs("set");
    }
  }

  if (options.nx && options.xx) {
    throw errWrongArgs("set");
  }
  if (options.keepTTL && hasExpiry) {
    throw errWrongArgs("set");
  }

  return options;
}

export function translateGet(args: string[]): Translation {
  if (args.length !== 1) {
    throw errWrongArgs("get");
  }

  const key = args[0];
  const statements: Statement[] = [
    deleteIfExpired(key),
    { sql: "SELECT type FROM keys WHERE key = ?", params: [key] },
    { sql: "SELECT value FROM strings WHERE key = ?", params: [key] },
  ];

  return {
    statements,
    mapResult: (results) => {
      if (hasWrongType(results, 1)) {
        throw ErrWrongType;
      }
      if (results[2].length === 0) {
        return null;
      }
      return rowString(results[2][0], "value");
    },
  };
}

export function translateSet(args: string[]): Translation {
  if (args.length < 2) {
    throw errWrongArgs("set");
  }

  const key = args[0];
  const value = args[1];
  const options = parseSetOptions(args.slice(2));

  let expiresAt: number | null = null;
  if (!options.keepTTL) {
    expiresAt = options.expiresAt ?? null;
  }

  const statements: Statement[] = [deleteIfExpired(key)];
  let getIndex = -1;
  if (options.get) {
    getIndex = statements.length;
    statements.push({ sql: "SELECT value FROM strings WHERE key = ?", params: [key] });
  }

  let okIndex = -1;
  if (options.nx) {
    okIndex = statements.length;
    statements.push(
      {
        sql: "INSERT OR IGNORE INTO keys (key, type, expires_at) VALUES (?, 'string', ?) RETURNING key",
        params: [key, expiresAt],
      },
      {
        sql: "INSERT INTO strings (key, value) SELECT ?, ? WHERE changes() > 0",
        params: [key, value],
      }
    );
  } else if (options.xx) {
    okIndex = statements.length;
    if (options.keepTTL) {
      statements.push({
        sql: "UPDATE keys SET type = 'string' WHERE key = ? RETURNING key",
        params: [key],
      });
    } else {
      statements.push({
        sql: "UPDATE keys SET type = 'string', expires_at = ? WHERE key = ? RETURNING key",
        params: [expiresAt, key],
      });
    }
    statements.push({
      sql: "INSERT OR REPLACE INTO strings (key, value) SELECT ?, ? WHERE changes() > 0",
      params: [key, value],
    });
  } else {
    if (options.keepTTL) {
      statements.push(
        {
          sql: "UPDATE keys SET type = 'string' WHERE key = ? RETURNING key",
          params: [key],
        },
        {
          sql: "INSERT INTO keys (key, type) SELECT ?, 'string' WHERE changes() = 0",
          params: [key],
        },
        {
          sql: "INSERT OR REPLACE INTO strings (key, value) VALUES (?, ?)",
          params: [key, value],
        }
      );
    } else {
      statements.push(
        { sql: "DELETE FROM keys WHERE key = ?", params: [key] },
        {
          sql: "INSERT INTO keys (key, type, expires_at) VALUES (?, 'string', ?)",
          params: [key, expiresAt],
        },
        {
          sql: "INSERT INTO strings (key, value) VALUES (?, ?)",
          params: [key, value],
        }
      );
    }
  }

  return {
    statements,
    mapResult: (results) => {
      let oldValue: unknown = null;
      if (getIndex >= 0) {
        if (results[getIndex].length === 0) {
          oldValue = null;
        } else {
          oldValue = rowString(results[getIndex][0], "value");
        }
      }

      if (okIndex >= 0 && results[okIndex].length === 0) {
        return null;
      }
      if (options.get) {
        return oldValue;
      }
      return "OK";
    },
  };
}

export function translateDel(args: string[]): Translation {
  if (args.length === 0) {
    throw errWrongArgs("del");
  }

  const statements = args.map((key) => ({
    sql: "DELETE FROM keys WHERE key = ? RETURNING key",
    params: [key],
  }));

  return {
    statements,
    mapResult: (results) => {
      let count = 0;
      for (const res of results) {
        count += res.length;
      }
      return count;
    },
  };
}

export function translateExists(args: string[]): Translation {
  if (args.length === 0) {
    throw errWrongArgs("exists");
  }

  const statements = args.map((key) => ({
    sql: "SELECT 1 FROM keys WHERE key = ? AND (expires_at IS NULL OR expires_at > unixepoch())",
    params: [key],
  }));

  return {
    statements,
    mapResult: (results) => {
      let count = 0;
      for (const res of results) {
        if (res.length > 0) {
          count++;
        }
      }
      return count;
    },
  };
}

function translateIncrByDelta(args: string[], delta: number, cmd: string): Translation {
  if (args.length !== 1) {
    throw errWrongArgs(cmd);
  }

  const key = args[0];
  const statements: Statement[] = [
    deleteIfExpired(key),
    { sql: "SELECT type FROM keys WHERE key = ?", params: [key] },
    { sql: "INSERT OR IGNORE INTO keys (key, type) VALUES (?, 'string')", params: [key] },
    { sql: "INSERT OR IGNORE INTO strings (key, value) VALUES (?, '0')", params: [key] },
    { sql: "SELECT value FROM strings WHERE key = ?", params: [key] },
    {
      sql: "UPDATE strings SET value = CAST(CAST(value AS INTEGER) + ? AS INTEGER) WHERE key = ? RETURNING value",
      params: [delta, key],
    },
  ];

  return {
    statements,
    mapResult: (results) => {
      if (hasWrongType(results, 1)) {
        throw ErrWrongType;
      }
      if (results[4].length === 0) {
        throw ErrNotInteger;
      }
      const before = rowString(results[4][0], "value");
      if (parseIntStrict(before) === null) {
        throw ErrNotInteger;
      }
      if (results[5].length === 0) {
        throw ErrNotInteger;
      }
      const after = rowString(results[5][0], "value");
      const value = parseIntStrict(after);
      if (value === null) {
        throw ErrNotInteger;
      }
      return value;
    },
  };
}

export function translateIncr(args: string[]): Translation {
  return translateIncrByDelta(args, 1, "incr");
}

export function translateIncrBy(args: string[]): Translation {
  if (args.length !== 2) {
    throw errWrongArgs("incrby");
  }
  const delta = parseIntStrict(args[1]);
  if (delta === null) {
    throw ErrNotInteger;
  }
  return translateIncrByDelta(args.slice(0, 1), delta, "incrby");
}

export function translateDecr(args: string[]): Translation {
  return translateIncrByDelta(args, -1, "decr");
}

export function translateDecrBy(args: string[]): Translation {
  if (args.length !== 2) {
    throw errWrongArgs("decrby");
  }
  const delta = parseIntStrict(args[1]);
  if (delta === null) {
    throw ErrNotInteger;
  }
  return translateIncrByDelta(args.slice(0, 1), -delta, "decrby");
}

export function translateMGet(args: string[]): Translation {
  if (args.length === 0) {
    throw errWrongArgs("mget");
  }

  const statements: Statement[] = [];
  for (const key of args) {
    statements.push(deleteIfExpired(key));
    statements.push({ sql: "SELECT value FROM strings WHERE key = ?", params: [key] });
  }

  return {
    statements,
    mapResult: (results) => {
      const out: unknown[] = [];
      for (let i = 0; i < args.length; i++) {
        const res = results[i * 2 + 1];
        if (res.length === 0) {
          out.push(null);
        } else {
          out.push(rowString(res[0], "value"));
        }
      }
      return out;
    },
  };
}

export function translateMSet(args: string[]): Translation {
  if (args.length === 0 || args.length % 2 !== 0) {
    throw errWrongArgs("mset");
  }

  const statements: Statement[] = [];
  for (let i = 0; i < args.length; i += 2) {
    const key = args[i];
    const value = args[i + 1];
    statements.push(
      { sql: "DELETE FROM keys WHERE key = ?", params: [key] },
      { sql: "INSERT INTO keys (key, type) VALUES (?, 'string')", params: [key] },
      { sql: "INSERT INTO strings (key, value) VALUES (?, ?)", params: [key, value] }
    );
  }

  return {
    statements,
    mapResult: () => "OK",
  };
}
